utils/src/bitmap/prunable.rs

```diff
@@ -0,0 +1,658 @@
+    ///
+    /// # Warning
+    ///
+    /// Panics if the bitmap is empty.
+    pub fn pop(&mut self) -> bool {
+        self.bitmap.pop()
+    }
+
+    /// Efficiently add a byte to the bitmap.
+    ///
+    /// # Warning
+    ///
+    /// Panics if self.next_bit is not byte aligned.
+    pub fn push_byte(&mut self, byte: u8) {
+        self.bitmap.push_byte(byte);
+    }
+
+    /// Efficiently add a chunk of bits to the bitmap.
+    ///
+    /// # Warning
+    ///
+    /// Panics if self.next_bit is not chunk aligned.
+    pub fn push_chunk(&mut self, chunk: &[u8; N]) {
+        self.bitmap.push_chunk(chunk);
+    }
+
+    /* Pruning */
+
+    /// Prune the bitmap to the most recent chunk boundary that contains the referenced bit.
+    ///
+    /// # Warning
+    ///
+    /// Panics if the referenced bit is greater than the number of bits in the bitmap.
+    pub fn prune_to_bit(&mut self, bit_offset: u64) {
+        let chunk_num = Self::raw_chunk_index(bit_offset);
+        if chunk_num < self.pruned_chunks {
+            return;
+        }
+
+        let chunks_to_prune = chunk_num - self.pruned_chunks;
+        self.bitmap.prune_chunks(chunks_to_prune);
+        self.pruned_chunks = chunk_num;
+    }
+
+    /* Indexing Helpers */
+
+    /// Convert a bit offset into a bitmask for the byte containing that bit.
+    #[inline]
+    pub fn chunk_byte_bitmask(bit_offset: u64) -> u8 {
+        BitMap::<N>::chunk_byte_bitmask(bit_offset)
+    }
+
+    /// Convert a bit offset into the offset of the byte within a chunk containing the bit.
+    #[inline]
+    pub fn chunk_byte_offset(bit_offset: u64) -> usize {
+        BitMap::<N>::chunk_byte_offset(bit_offset)
+    }
+
+    /// Convert a bit offset into the index of the chunk it belongs to within the bitmap,
+    /// taking pruned chunks into account. That is, the returned value is a valid index into
+    /// the inner bitmap.
+    ///
+    /// # Warning
+    ///
+    /// Panics if the bit doesn't exist or has been pruned.
+    #[inline]
+    pub fn pruned_chunk_index(&self, bit_offset: u64) -> usize {
+        assert!(bit_offset < self.len(), "out of bounds: {bit_offset}");
+        let chunk_num = Self::raw_chunk_index(bit_offset);
+        assert!(chunk_num >= self.pruned_chunks, "bit pruned: {bit_offset}");
+
+        chunk_num - self.pruned_chunks
+    }
+
+    /// Convert a bit offset into the number of the chunk it belongs to,
+    /// ignoring any pruning.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `bit_offset / CHUNK_SIZE_BITS > usize::MAX`.
+    #[inline]
+    pub fn raw_chunk_index(bit_offset: u64) -> usize {
+        BitMap::<N>::chunk_index(bit_offset)
+    }
+
+    /// Get a reference to a chunk by its index in the current bitmap
+    /// Note this is an index into the chunks, not a bit offset.
+    #[inline]
+    pub fn get_chunk_by_index(&self, index: usize) -> &[u8; N] {
+        self.bitmap.get_chunk_by_index(index)
+    }
+}
```

Comment:
Should we implement `Read/Write` on this?

While we may not need it atm, it still seems useful (we'd encode `pruned_chunks` as `u64` I suppose?)